libraries/chain/block_log.cpp

```diff
@@ -899,11 +899,21 @@ namespace eosio { namespace chain {
          EOS_ASSERT(entry.block_num() == block_num, block_log_exception,
                      "Wrong block was read from block log.");
 
-         auto pruner =
-             overloaded{[](transaction_id_type&) { return false; },
-                        [&ids](packed_transaction& ptx) {
-                           return std::find(ids.begin(), ids.end(), ptx.id()) != ids.end() && (ptx.prune_all(), true);
-                        }};
+         auto pruner = overloaded{[](transaction_id_type&) { return false; },
+                                  [&ids](packed_transaction& ptx) {
+                                     auto it = std::find(ids.begin(), ids.end(), ptx.id());
+                                     if (it != ids.end()) {
+                                        ptx.prune_all();
+                                        // remove the found entry from ids
+                                        if (it != ids.end() - 1) {
+                                           *it = ids.back();
+                                        }
+                                        ids.resize(ids.size() - 1);
```

Comment:
I think this would be clearer as `ids.erase(it);` and the comment is not even needed then.

Reply:
The complexity of ids.erase(it) is O(n) while it is O(1) in current implementation.

Reply:
Understood. Personally I think it is worth the cost here for clearer code.

Reply:
done